src/websock/native/integrate_test.go

```diff
@@ -0,0 +1,508 @@
+				dead:  end,
+				wIn:   make(chan []byte),
+			}
+
+			go func() {
+				defer GinkgoRecover()
+
+				<-start
+
+				err := subject.Handle(websocket, req)
+				log.Println("got", err.Error(), ", handler closed")
+			}()
+		})
+
+		AfterEach(func() {
+			websocket = nil
+			close(end)
+		})
+
+		It("limits a call to the server by the client timeout when the servers' timeout is longer", func() {
+
+			deadline := make(chan time.Time)
+
+			peer.Listen(ns, func(ctx context.Context, req rinq.Request, res rinq.Response) {
+				// never return anything
+
+				dead, _ := ctx.Deadline()
+				deadline <- dead
+			})
+
+			// client timeouts are in milliseconds
+			clientTimeout := time.Duration(2000)
+			serverTimeout := 10 * time.Second
+
+			subject = native.NewHandler(peer, message.JSONEncoding, native.MaxCallTimeout(serverTimeout))
+
+			websocket.queueMsg(createSession, session, nil, nil)
+			websocket.queueMsg(callSync, session, []interface{}{
+				seq, ns, cmd, clientTimeout,
+			}, "ping")
+
+			expectedTime := time.Now().Add(clientTimeout * time.Millisecond)
+
+			close(start)
+
+			Expect(<-deadline).To(BeTemporally("~", expectedTime, time.Second/2))
+		})
+
+		It("limits a call to the server by the server timeout when the clients' timeout is longer", func() {
+			deadline := make(chan time.Time)
+
+			peer.Listen(ns, func(ctx context.Context, req rinq.Request, res rinq.Response) {
+				// never return anything
+
+				dead, _ := ctx.Deadline()
+				deadline <- dead
+			})
+
+			// client timeouts are in milliseconds
+			clientTimeout := time.Duration(10000)
+			serverTimeout := time.Duration(2000) * time.Millisecond
+
+			subject = native.NewHandler(peer, message.JSONEncoding, native.MaxCallTimeout(serverTimeout))
+
+			websocket.queueMsg(createSession, session, nil, nil)
+			websocket.queueMsg(callSync, session, []interface{}{
+				seq, ns, cmd, clientTimeout,
+			}, "ping")
+
+			expectedTime := time.Now().Add(serverTimeout)
+
+			close(start)
+
+			Expect(<-deadline).To(BeTemporally("~", expectedTime, time.Second/2))
+		})
+
+		It("limits a call to the server by the clients' timeout when the servers' timeout is not set", func() {
+			deadline := make(chan time.Time)
+
+			peer.Listen(ns, func(ctx context.Context, req rinq.Request, res rinq.Response) {
+				// never return anything
+
+				dead, _ := ctx.Deadline()
+				deadline <- dead
+			})
+
+			// client timeouts are in milliseconds
+			clientTimeout := time.Duration(10000)
+
+			subject = native.NewHandler(peer, message.JSONEncoding)
+
+			websocket.queueMsg(createSession, session, nil, nil)
+			websocket.queueMsg(callSync, session, []interface{}{
+				seq, ns, cmd, clientTimeout,
+			}, "ping")
+
+			expectedTime := time.Now().Add(clientTimeout * time.Millisecond)
+
+			close(start)
+
+			Expect(<-deadline).To(BeTemporally("~", expectedTime, time.Second/2))
+		})
+
+	})
+})
+
+func msg(msgType uint16, session uint16, headers interface{}, payload interface{}) []byte {
```

Comment:
It's not clear what this is for. When the test are abstracted heavily in this way, please name document code as you would for non-test code.

I think this could be named `encodeFrame()` or something like that to make it clearer.

Reply:
I've flicked it over to be `encodeClientReq` so it feels wrong to be deserialising server responses with it